core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java

```diff
@@ -0,0 +1,634 @@
+            }
+            return valueFuncs.getQuick(col);
+        }
+
+        private void initBounds(Function fromFunc, Function toFunc) {
+            if (fromFunc != TimestampConstant.NULL && toFunc != TimestampConstant.NULL) {
+                rangeBound = RANGE_FULLY_BOUND;
+                return;
+            }
+
+            if (fromFunc != TimestampConstant.NULL) {
+                rangeBound = RANGE_LOWER_BOUND;
+                return;
+            }
+
+            if (toFunc != TimestampConstant.NULL) {
+                rangeBound = RANGE_UPPER_BOUND;
+                return;
+            }
+
+            rangeBound = RANGE_UNBOUNDED;
+        }
+
+        private void initTimestamps(Function fromFunc, Function toFunc, CharSequence stride) throws SqlException {
+            if (fromFunc != TimestampConstant.NULL) {
+                fromTimestamp = fromFunc.getTimestamp(null);
+            }
+
+            if (toFunc != TimestampConstant.NULL) {
+                toTimestamp = toFunc.getTimestamp(null);
+            }
+
+            timestampSampler = TimestampSamplerFactory.getInstance(stride, 0);
+            timestampSampler.setStart(fromTimestamp);
+
+            nextBucketTimestamp = fromTimestamp;
+
+            minTimestamp = fromFunc == TimestampConstant.NULL ? Long.MAX_VALUE : fromTimestamp;
+            maxTimestamp = toFunc == TimestampConstant.NULL ? Long.MIN_VALUE : toTimestamp;
+        }
+
+        private void moveToNextBucket() {
+            bucketIndex++;
+            nextBucketTimestamp = timestampSampler.nextTimestamp(nextBucketTimestamp);
+        }
+
+        private boolean notAtEndOfBitset() {
+            if (rangeBound == RANGE_LOWER_BOUND || rangeBound == RANGE_UNBOUNDED) {
+                return bucketIndex < timestampSampler.bucketIndex(maxTimestamp);
+            } else {
+                return bucketIndex <= timestampSampler.bucketIndex(maxTimestamp);
+            }
+        }
+
+        private void of(
+                RecordCursor baseCursor,
+                @NotNull Function fromFunc,
+                @NotNull Function toFunc,
+                CharSequence stride,
+                ObjList<Function> valueFuncs,
+                int timestampIndex,
+                SqlExecutionContext executionContext
+        ) throws SqlException {
+            this.baseCursor = baseCursor;
+            this.timestampIndex = timestampIndex;
+            this.valueFuncs = valueFuncs;
+            Function.init(valueFuncs, baseCursor, executionContext);
+            fromFunc.init(baseCursor, executionContext);
+            toFunc.init(baseCursor, executionContext);
+            initTimestamps(fromFunc, toFunc, stride);
+            if (presentRecords == null) {
+                presentRecords = new BitSet(toFunc != TimestampConstant.NULL ? timestampSampler.bucketIndex(toTimestamp) : DEFAULT_BITSET_SIZE);
+            }
+            valueFuncs.insert(timestampIndex, 1, null);
```

Comment:
This insertion doesn't seem to be idempotent and we're making this call on each query execution.